lib/erl_docgen/priv/bin/chunk.escript

```diff
@@ -0,0 +1,747 @@
+                                     {signature,Signature},
+                                     {meta,SinceMD#{ signature => Specs }}],
+                           ContentsNoName}
+                  end, lists:usort(FAs));
+            NameList ->
+                %% Manual style function docs
+                FAs = lists:flatten([func_to_tuple(NameString) || {name, _Attr, NameString} <- NameList]),
+
+                VerifyNameList(NameList,fun([_|_]) -> ok end),
+
+                Signature = [strip_tags(NameString) || {name, _Attr, NameString} <- NameList],
+                [{function,[{name,F},{arity,A},
+                            {signature,Signature},
+                            {meta,SinceMD}],ContentsNoName}
+                 || {F,A} <- lists:usort(FAs)]
+        end,
+    transform(Functions,[]).
+
+func_to_tuple(Chars) ->
+    try
+        [Name,Args] = string:split(strip_tags(Chars),"("),
+        Arities = parse_args(unicode:characters_to_list(Args)),
+        [{unicode:characters_to_list(Name),Arity} || Arity <- Arities]
+    catch E:R:ST ->
+            io:format("Failed to parse: ~p~n",[Chars]),
+            erlang:raise(E,R,ST)
+    end.
+
+%% This function parses a documentation <name> attribute to figure
+%% out the arities if that function. Example:
+%%    "start([go,Mode] [,Extra])" returns [1, 2].
+%%
+%% This assumes that when a single <name> describes many arities
+%% the arities are listed with [, syntax.
+parse_args(")" ++ _) ->
+    [0];
+parse_args(Args) ->
+    parse_args(unicode:characters_to_list(Args),1,[]).
+parse_args([$[,$,|T],Arity,[]) ->
+    parse_args(T,Arity,[$[]) ++ parse_args(T,Arity+1,[]);
+parse_args([$,|T],Arity,[]) ->
+    parse_args(T,Arity+1,[]);
+parse_args([Open|T],Arity,Stack)
+  when Open =:= $[; Open =:= ${; Open =:= $( ->
+    parse_args(T,Arity,[Open|Stack]);
+parse_args([$]|T],Arity,[$[|Stack]) ->
+    parse_args(T,Arity,Stack);
+parse_args([$}|T],Arity,[${|Stack]) ->
+    parse_args(T,Arity,Stack);
+parse_args([$)|T],Arity,[$(|Stack]) ->
+    parse_args(T,Arity,Stack);
+parse_args([$)|_T],Arity,[]) ->
+    [Arity];
+parse_args([_H|T],Arity,Stack) ->
+    parse_args(T,Arity,Stack).
+
+strip_tags([{_Tag,_Attr,Content}|T]) ->
+    [Content | strip_tags(T)];
+strip_tags([H|T]) when not is_tuple(H) ->
+    [H | strip_tags(T)];
+strip_tags([]) ->
+    [].
+
+transform_datatype(Dom,_Acc) ->
+    ContentsNoName = transform([NC||NC <- Dom, element(1,NC) /= name],[]),
+    [case N of
+          {name,NameAttr,[]} ->
+              {datatype,NameAttr,ContentsNoName};
+          {name,[],Content} ->
+              [{Name,Arity}] = func_to_tuple(Content),
+              Signature = strip_tags(Content),
+              {datatype,[{name,Name},{n_vars,integer_to_list(Arity)},
+                         {signature,Signature}],ContentsNoName}
+      end || N = {name,_,_} <- Dom].
+
+transform_seealso(_S = {seealso,_Attr,_Content}) ->
+    _Content.
```

Comment:
here are couple examples of `seealso` in action in xml:

* `<p>See also <seealso marker="#new-2"><c>new/2</c></seealso>`
* `<seealso marker="kernel:file"><c>file</c></seealso> module.</p>`

currently we transform this into `{c, [], ["new/2"]}` and `{c, [], ["file"]}` respectively. This is a bit broader discussion and probably not a big priority at the moment, but currently we lose semantic information which is especially relevant for the `file` atom: was it suppose to mean just an _atom_ or a _module_? To preserve references, perhaps this should be transformed into something like:

* `{a, [{href, "#new-2"}], [{c, [], "new/2"}]}`
* `{a, [{href, "kernel:file"}], [{c, [], "file"}]}`

and then the downstream tools could interpret these to create proper cross-references. To differentiate from external URLs (we sometimes link to e.g. RFCs) the `a` element could have additional attributes.

Reply:
EEP 48 supports cross references in metadata, so we could move those there. This may even allow us to remove paragraphs such as above completely and instead automatically generate them.

Reply:
So the more complete snippet from `filelib.xml` is:

```
  <description ghlink="maint/lib/stdlib/doc/src/filelib.xml#L38">
    <p>This module contains utilities on a higher level than the
      <seealso marker="kernel:file"><c>file</c></seealso> module.</p>
```

is the idea that this would be converted to:

```html
<p>This module contains utilities on a higher level than the
  <code>file</code>module.</p>

See also: <a href="kernel:file"><code>file</code></a>, ...
```

that is, we wouldn't have "inline" cross-references but they'd always happen and the end of a given block of docs?

Reply:
I think you would want both. Sometimes you want to provide some information when linking, but sometimes a direct “see also”, like in Man pages, is desired.

Reply:
I've not really looked at any of the linking yet as it is not needed for the shell docs. I don't have time to look at it now so it will have to wait until after OTP 23-rc1.

Reply:
Having re-read what you wrote.

Yes, `seealso` should be translated to a `<a>` tag. The reason why it is not done right now is because it was not needed for the shell documentation. Also, we do not want to do a direct mapping of what we have in `marker` to what is in the `<a>` as some of the information is unnecessarily hard to figure out later on.

In order to get something working at all with links I'll include `<a marker="kernel:file">` in our chunks for now. But don't do too much work on parsing these yet as they are bound to change in OTP 23-rc2.

Reply:
Sounds good, thank you!